Feature substring repo search

nadavsteindler opened the pull request:
Closes #7596

## Change Description

### Background

Community feature request 
      
### New Feature

Repo search will seek substrings rather than prefixes, making it more likely to turn up what the user wants

### Testing Details

How were the changes tested?

### Breaking Change?

Does this change break any existing functionality? (API, CLI, Clients)

## Additional info

Logs, outputs, screenshots of changes if applicable (CLI / GUI changes)

### Contact Details

How can we get in touch with you if we need more info? (ex. email@example.com)

arielshaqed commented on pkg/catalog/catalog.go:
```diff
@@ -598,36 +598,25 @@ func (c *Catalog) ListRepositories(ctx context.Context, limit int, prefix, after
-		startPos = afterRepositoryID
-	}
-	if startPos != "" {
-		it.SeekGE(startPos)
```
This makes scanning _very_ inefficient.

nadavsteindler replied:
Sorry I fixed it- seek to start position is back in

guy-har replied:
But now we are missing repositories that include the string and are before the prefix (e.g for prefix ple we will never get apple)

nadavsteindler replied:
OK I implemented the substring part. I thought to just call it search string, since that is what the client is trying to do and maybe we will improve the algorithm at some point- let me know what you think

arielshaqed commented on pkg/catalog/catalog.go:
```diff
@@ -598,36 +598,25 @@ func (c *Catalog) ListRepositories(ctx context.Context, limit int, prefix, after
-		// collect limit +1 to return limit and has more
-		if len(repos) >= limit+1 {
-			break
+		if strings.Contains(string(record.RepositoryID), prefix) {
```
This is a breaking change: it changes the meaning of the "prefix" parameter!  (It is also very confusing to users, who will probably not expect "prefix" to mean "substring").

nadavsteindler replied:
I guess I could change the name of the parameter- I don't see any other code that uses it

guy-har replied:
Prefix is defined in our API and should still be supported, even if we wanted, we can't just change it.

nadavsteindler replied:
So we have to support prefix and also a search string?

arielshaqed commented on pkg/catalog/catalog.go:
```diff
@@ -598,36 +598,25 @@ func (c *Catalog) ListRepositories(ctx context.Context, limit int, prefix, after
-	prefixRepositoryID := graveler.RepositoryID(prefix)
-	startPos := prefixRepositoryID
-	if afterRepositoryID > startPos {
-		startPos = afterRepositoryID
```
Doesn't this break pagination?

nadavsteindler replied:
Yes you are correct thanks. Added some failing tests and fixed it

arielshaqed reviewed (changes requested):
This feature worries me _a lot_.

## Cost & performance

Suppose a user has 100 repositories (any less and they probably don't need it).  They start typing a repository name on the webUI.  Every character they type on the GUI narrows the search, at the cost of a KV listing.  A reasonable typist can hit 10 characters per second; didn't we just hit 1000 RCUs/s?

Ideally we would open an immediate tech debt item to improve KV to allow faster scanning: we could add conditions on all supported KVs that would greatly reduce the required traffic.

Also, existing programs that perform list-repositories may now cost much more: _every repository listing scans all repositories_.

## API

AFAICT the code overrides "prefix" to mean "substring".  This is both confusing and a breaking change.  At the very least, please keep "prefix", and add a new "substring" field.

Also, I believe that this change breaks pagination; please test and fix if needed.

nadavsteindler commented:
> This feature worries me _a lot_.
> 
> ## Cost & performance
> Suppose a user has 100 repositories (any less and they probably don't need it). They start typing a repository name on the webUI. Every character they type on the GUI narrows the search, at the cost of a KV listing. A reasonable typist can hit 10 characters per second; didn't we just hit 1000 RCUs/s?
> 
> Ideally we would open an immediate tech debt item to improve KV to allow faster scanning: we could add conditions on all supported KVs that would greatly reduce the required traffic.
> 
> Also, existing programs that perform list-repositories may now cost much more: _every repository listing scans all repositories_.
> 
> ## API
> AFAICT the code overrides "prefix" to mean "substring". This is both confusing and a breaking change. At the very least, please keep "prefix", and add a new "substring" field.
> 
> Also, I believe that this change breaks pagination; please test and fix if needed.

OK OK, let's not get over-excited. I deleted a few lines too much, but now I fixed them, so I think it should be fine. I can performance test it to make sure

guy-har commented:
@arielshaqed, can you please explain the calculation that brings us to  1000 RCUs/s? 
AFAIU a scan costs 1RCU for 4KB of data, so roughly speaking scanning 100 records would be around 4 RCUs

arielshaqed commented:
> @arielshaqed, can you please explain the calculation that brings us to 1000 RCUs/s? AFAIU a scan costs 1RCU for 4KB of data, so roughly speaking scanning 100 records would be around 4 RCUs

Sorry, you're right.  I always make the same mistake: "kv".Store.Scan is actually DynamoDB Query, which indeed costs 1 RCU per 4 KiB of data.  How large is a repository record?
```pb
message RepositoryData {
  string id = 1;
  string storage_namespace = 2;
  string default_branch_id = 3;
  google.protobuf.Timestamp creation_date = 4;
  RepositoryState state = 5;
  string instance_uid = 6;
  bool read_only = 7;
}
```
Recalculating: I can certainly see >160 bytes per repository, which is indeed just under 25 repositories per RCU, so 4-5 RCUs per keystroke.  Obviously not ideal but please strike my original numbers (and I'm editing that comment to clarify the error).

nadavsteindler commented on pkg/catalog/catalog_test.go:
```diff
@@ -75,10 +75,12 @@ func TestCatalog_ListRepositories(t *testing.T) {
 		{RepositoryID: "repo1", Repository: &graveler.Repository{StorageNamespace: "storage1", CreationDate: now, DefaultBranchID: "main1"}},
 		{RepositoryID: "repo2", Repository: &graveler.Repository{StorageNamespace: "storage2", CreationDate: now, DefaultBranchID: "main2"}},
 		{RepositoryID: "repo3", Repository: &graveler.Repository{StorageNamespace: "storage3", CreationDate: now, DefaultBranchID: "main3"}},
+		{RepositoryID: "repo22", Repository: &graveler.Repository{StorageNamespace: "storage4", CreationDate: now, DefaultBranchID: "main4"}},
```
Is there an assumption that they are inserted in order?

arielshaqed replied:
I don't see how there could be.  However the test _can_ and _should_ verify that they are returned in order, because that's the API.

nadavsteindler commented:
> probably don't need it). They start typing a repository name on the webUI. Every character they type on the GUI narrows the search, at the cost of a KV listing. A reasonable typist can hit 10 characters per second; didn't we just hit 1000 RCUs/s?

@itaigilo pointed out to me that our frontend uses useDebouncedState with a wait time of 300ms, so even if the user types faster, it will generate at most one request per 300ms

itaigilo commented on api/swagger.yml:
```diff
@@ -70,6 +70,13 @@ components:
+
+    PaginationSearchString:
+      in: query
+      name: search
```
I think the name `search` is too vague.
Maybe `query` or `search-query` is more accurate in this context.

Moreover, I find it weird that a `list` API is getting some searching capabilities as a main part of its signature.
Maybe a structure containing some search params might be more clear, but it's tricky when using a GET request.

How about the following solution:
Supporting `*<search-term>*` pattern for the `prefix` param.
When the API encounters such pattern, it performs a substring search, and it's not really breaking the API.
And the WebUI can wrap the `prefix` with `*`s, and actually make such search.

nadavsteindler replied:
I got rid of PaginationSearchTerm- I agree that is just too much

arielshaqed replied:
> I got rid of PaginationSearchTerm- I agree that is just too much

Yeah, it's not part of pagination.

itaigilo commented on pkg/catalog/catalog.go:
```diff
@@ -615,19 +615,21 @@ func (c *Catalog) ListRepositories(ctx context.Context, limit int, prefix, after
-		// collect limit +1 to return limit and has more
-		if len(repos) >= limit+1 {
-			break
+		if strings.Contains(string(record.RepositoryID), searchString) {
```
I think it would make the code cleaner if it was:
```
if !strings.Contains(string(record.RepositoryID), searchString) {
  continue
}
```

itaigilo reviewed (changes requested):
Thanks for the effort, this is a tricky full-stack feature.

Blocking for both the param naming,
And for at least considering a different design of the API.

nadavsteindler commented on api/swagger.yml:
```diff
@@ -70,6 +70,13 @@ components:
       schema:
         type: string
+
+    PaginationSearchString:
```
Hmm, well a "search string" is a term: https://www.techtarget.com/whatis/definition/search-string
The typically query strings are shortened to one word. See for example:
![image](https://github.com/user-attachments/assets/539f126c-12cf-4480-96cf-93f0d010969c)

nadavsteindler commented on api/swagger.yml:
```diff
@@ -695,7 +702,7 @@ components:
         hidden:
           type: boolean
-          description: When set, branch will not show up when listing branches by default. *EXPERIMENTAL*          
+          description: When set, branch will not show up when listing branches by default. *EXPERIMENTAL*
```
Do you know how? I just spent half an hour at this and it's not working.
I turned off code formatting on autosave- it erases the changes
I turned off go fmt on commit- it erases the changes
I tried adding the changes from the command line- it erases the changes
I tried editing the file in the github UI- it has different linebreaks
Not sure it's worth the time. In general we shouldn't be committing all these extra spaces...

nadavsteindler replied:
The formatting in this file is anyway inconsistent- I think we need to format it once and then never have to think about it again

itaigilo replied:
Alright, I'm good with formatting this one once.
Not sure how we can enforce this in the future - 
But anyway, it's worth the time imho.

nadavsteindler commented:
> Thanks for the effort, this is a tricky full-stack feature.
> 
> Blocking for both the param naming, And for at least considering a different design of the API.

OK I addressed these issues (see the comments)

arielshaqed commented on pkg/api/controller.go:
```diff
@@ -5470,6 +5470,13 @@ func paginationDelimiter(v *apigen.PaginationDelimiter) string {
 	return string(*v)
 }
 
+func paginationSearch(v *apigen.SearchString) string {
```
```suggestion
func search(v *apigen.SearchString) string {
```
I don't see what this has to do with pagination.

arielshaqed commented on webui/src/lib/api/index.js:
```diff
@@ -458,8 +458,9 @@ class Repositories {
-    async list(prefix = "", after = "", amount = DEFAULT_LISTING_AMOUNT) {
-        const query = qs({prefix, after, amount});
+    async list(search = "", after = "", amount = DEFAULT_LISTING_AMOUNT) {
+        console.log("list search ", search)
```
Remove development leftovers:
```suggestion
```

arielshaqed reviewed (approved):
Neat, thanks!  Please see comments about it being a _search_ string, not a _pagination_ search string.  Now we'll see if anyone was depending on the old behaviour.[^1]

[^1]:  See [XKCD #1172](https://xkcd.com/1172/).

itaigilo commented on webui/src/pages/repositories/index.jsx:
```diff
@@ -190,10 +190,10 @@ const RepositoriesPage = () => {
-    const routerPfx = (router.query.prefix) ? router.query.prefix : "";
-    const [prefix, setPrefix] = useDebouncedState(
+    const routerPfx = (router.query.search) ? router.query.search : "";
+    const [search, setsearch] = useDebouncedState(
```
Should be `setSearch`.

itaigilo reviewed (approved):
LGTM, nice one.

Approving to unblock,
But please fix Ariel's and my style comments before merging (`paginationSearch`, `setsearch`, and `console.log`).

guy-har commented on pkg/catalog/catalog.go:
```diff
@@ -585,7 +585,7 @@ func (c *Catalog) DeleteRepositoryMetadata(ctx context.Context, repository strin
 
 // ListRepositories list repository information, the bool returned is true when more repositories can be listed.
 // In this case, pass the last repository name as 'after' on the next call to ListRepositories
-func (c *Catalog) ListRepositories(ctx context.Context, limit int, prefix, after string) ([]*Repository, bool, error) {
+func (c *Catalog) ListRepositories(ctx context.Context, limit int, prefix, searchString, after string) ([]*Repository, bool, error) {
```
Document searchString
